kernel/mmanutils.c

```diff
@@ -99,14 +107,48 @@ static ssize_t _map_file_onto_memory(
 
             memcpy(p, buf, (size_t)n);
             p += n;
+            o += n;
             r -= (size_t)n;
             bytes_read += n;
         }
     }
 
-    /* restore the file position */
-    if (lseek(fd, save_pos, SEEK_SET) == (off_t)-1)
-        goto done;
+    /* get the fd flags */
+    ECHECK(flags = fcntl(fd, F_GETFL));
+
+    /* if file is writable, then create msync mappings for msync() */
+    if ((mmap_flags & MAP_SHARED) && flags & (O_RDWR | O_WRONLY))
+    {
+        const size_t n = length / PAGE_SIZE;
+        size_t o = offset;
+        uint8_t* p = addr;
+
+        myst_spin_lock(&_msync_mappings_lock);
+
+        for (size_t i = 0; i < n; i++)
+        {
+            msync_mapping_t* fm;
+
+            if (!(fm = calloc(1, sizeof(msync_mapping_t))))
+            {
+                myst_spin_unlock(&_msync_mappings_lock);
+                ERAISE(-ENOMEM);
+            }
+
+            fm->fd = fd;
```

Comment:
It seems inefficient to create a msync_mapping_t entry for every page. Can we optimize?

Reply:
I updated the PR to fix this. Now msync mappings cover the entire range and are reduced or split when unmappings occur.